Foundation/URLSession/http/EasyHandle.swift

```diff
@@ -172,20 +169,20 @@ extension _EasyHandle {
         let protocols = (CFURLSessionProtocolHTTP | CFURLSessionProtocolHTTPS)
         try! CFURLSession_easy_setopt_long(rawHandle, CFURLSessionOptionPROTOCOLS, protocols).asError()
         try! CFURLSession_easy_setopt_long(rawHandle, CFURLSessionOptionREDIR_PROTOCOLS, protocols).asError()
-        #if os(Android)
-            // See https://curl.haxx.se/docs/sslcerts.html
-            // For SSL to work you need "cacert.pem" to be accessable
-            // at the path pointed to by the URLSessionCAInfo env var.
-            // Downloadable here: https://curl.haxx.se/ca/cacert.pem
-            if let caInfo = _EasyHandle._CAInfoFile  {
-                if String(cString: caInfo) == "UNSAFE_SSL_NOVERIFY" {
-                    try! CFURLSession_easy_setopt_int(rawHandle, CFURLSessionOptionSSL_VERIFYPEER, 0).asError()
-                }
-                else {
-                    try! CFURLSession_easy_setopt_ptr(rawHandle, CFURLSessionOptionCAINFO, caInfo).asError()
-                }
+#if os(Android)
+        // See https://curl.haxx.se/docs/sslcerts.html
+        // For SSL on Android you need a "cacert.pem" to be
+        // accessible at the path pointed to by this env var.
+        // Downloadable here: https://curl.haxx.se/ca/cacert.pem
+        if let caInfo = getenv("URLSessionCertificateAuthorityInfoFile")  {
```

Comment:
Thanks for making this change, I was able to get SSL working on Android today using this env variable. Is this really the nicest way about this though? The fact that it's alterable like this is kind of confusing as an API consumer, plus it looks a lot like a security risk to me.

https://curl.haxx.se/docs/sslcerts.html details the option of building libcurl with the flag `--with-ca-bundle=FILE`. So from what I understand, we could remove this code complexity entirely by just building libcurl with the certificate mentioned in your comment here. What do you think @johnno1962 ?

Reply:
Hi @ephemer, there was a lot of to-ing and fro-ing on the PR. The original PR #1103 which was merged was an API which was developed in #1113 until it was decided we couldn't add api to foundation and we switched back to the original proposal. As far as one can know I don’t think it’s a security issue as if you use it it will be set explicitly in the code and there will be no opportunity to root the device and override it by setting the var maliciously for example. I wasn’t aware you could hard code it into libcurl but that doesn’t help as it wouldn't have a fixed path on the menagerie that is android.

If I were to do it all again and known #1113 wouldn’t be merged for three months. I would have left it as it was after #1103 but by then the toolchain was committed to an environment variable.

Reply:
Hi @johnno1962, if I understand correctly, this is essentially a process-wide `public` variable that can be changed at any time before any `URLSession` is created. That certainly seems like a security risk to me. I don't know enough about SSL (or the finer details of Android for that matter) to know to what extent this may be risky, but it'd be hard to argue that it's secure. That is, unless you're calling `setenv` religiously before every initialization of `URLSession`, but that seems like an unreasonable expectation to have of Foundation's API consumers.

From what I understand, implementing `--with-ca-bundle` wouldn't mean hard-coding a file path into libcurl, but rather incorporating the certificate bundle into libcurl.so statically. If that's not the case I agree it'd be pointless though. I'll do some more testing on this tomorrow. If a static solution does turn out to be possible, would you be ok with this code being removed?

Reply:
Having fought hard to get this small change in I’d rather it wasn’t rolled back! Besides, I don’t see the alternative though I’m open to practical alternatives. You can hard code the path to the certificates in libcurl but the problem is since liibssl 1.0+ they are in the wrong format on most android devices, see http://www.javacms.tech/questions/1728116/how-to-make-ssl-peer-verify-work-on-android. Including the certificate bundle in libcurl i.e. the toolchain seems like a very bad idea to me.

Whether this presents a security risk is something I’d rather not express an opinion on but I doubt it. If you can’t relay on an apps code bundle all bets are probably off anyway. If not, using an environment variable or an api doesn’t make much difference. If you find a better way though then all power to your elbow.

Reply:
Hi @johnno1962, thanks for your patience. I've just spent a couple of hours reading into this and have come to the conclusion that you may well have the only reasonable solution that is usable for this. I am still sceptical that it's entirely secure but now equally as sceptical that there's any other way on Android.

Do you have an idea of how best to use this in an APK? libcurl expects an absolute path.

Reply:
Hi @ephemer, checkout https://github.com/SwiftJava/swift-android-kotlin for examples on how it all comes together. You need to have the cacerts file as a resource, copy it to the app’s cacheDir and set the environment variable to that path in Swift. There just didn’t seem to be a better way to do it.

Reply:
Thanks for the info. Ok, so you're basically copying the `.pem` from the APK's resources into the app cache (which has an absolute path), accessed via the applicationContext. It's a pain in the butt but I guess there's not much else we could do other than loading the `.pem` ourselves and adding the certs manually.

Maybe it's worth us putting in an issue to the NDK asking them to provide a libcurl (with SSL, using the system certificates)? It might take a few years, but hey..